### DISCUSSION TITLE
Mirror permissions on backend DB or openFGA as single source of truth

### DISCUSSION THREAD

**Discussion by Raid55** [openfga/community#241] [Q&A] [answered]
Is it recommended to keep a "mirror" of the permissions in our backend DB, such as groups, guests, members, their roles, etc... 
Existing in our DB as well as tuples in openFGA's DB.
Or is it better practice to keep openFGA as a single source of truth and to get info as to who has permissions on a specific document by querying openFGA (with ListObject, ListRelation, and Read calls).
For example:
We have a group table in the database to represent user groups and the groups display name and avatar. But now with openFGA, do we also need a group_member_user join table? Would we just be able to query openFGA to get the members of a specific group or is that and incorrect way to use openFGA?

To expand on this, this is mostly a UI question. When rendering a UI in the frontend we need to query the backend that will return the users who have access to the document (and what role they have). Should that information be stored in the backend database as join tables or is it "safe" and good practice to just call:
```
const guests = client.read({	 
    "tuple_key": {
        "relation": "guest",
        "object": "workspace:acme"
    }
})
const members = client.read({	 
    "tuple_key": {
        "relation": "member",
        "object": "workspace:acme"
    }
})
const admins = client.read({	 
    "tuple_key": {
        "relation": "admin",
        "object": "workspace:acme"
    }
})
```
 and then query our database by filtering the ID's in a sql query.
 
 The alternative is storing those relations such as (with typeorm) we can call `await workspace.members`/`await workspace.guests` and it will join the users.
 This would mean that every time we change a permission we also need to change that relation in our backend.
 
 For reference I read: https://openfga.dev/docs/interacting/search-with-permissions
 but where this differs is that i am not trying to check for access on a resource, i am trying to determine the level of access on each resource to return to the user and render them "share modal" UI along with their permission level.

**Comment by Raid55** [comment, marked as answer]
[@xtc on discord](https://discord.com/channels/759188666072825867/1196206916528001054/1196222359552663602) left some useful points to take into consideration and I believe it answers my question.

> some thoughts on this:
> 
> - How comfortable do you feel coupling your data relationships with OpenFGA? If you ever need to consider moving away form OpenFGA, you would need to either remodel your data or replicate the current source of truth (if possible) in the new service.
> 
> - When your applications scales, how do you ensure you're able to differentiate the resources required for your authz service and your "business data" service?
> 
> - It's generally best practise to have a separate DB for your OpenFGA, which means you'll likely need both services to have a complete picture of your data. Consider a scenario where you need to give a BI team access to this data,  this means your BI tooling would require access to both services / databases.
> 
> - You should only be storing information that is relevant for the purpose of authz, could using OpenFGA as the source of truth for some relations mean you might end up having to introduce data that would no longer be relevant?
> 
> - How would this impact performance compared to having your data in the one place where you may be able to easily retrieve these with a join? Maybe something to benchmark vs an extra round trip to OpenFGA.
> 
> - How does his impact your DR strategies? You would need to ensure your backups are now linked or can be linked during recovery, potentially impacting Recovery Point Objective (RPO).
> 
> My personal opinion:  I don't think this would be good practise. I would keep your authz layer (OpenFGA) and your business data layer decoupled. They are doing fundamentally different things and just because OpenFGA is providing some endpoints to do this, using it as the source of truth for potentially unrelated things (i.e. rendering UI) may not be its intended use.

**Reply by Sagensagen** [reply]
I am in this scenario where i keep all the groups, roles, functionality within our relational db and can simply mirror them into OpenFGA just the way described here. What i am struggling with is how to maintain a synchronized system. It is possible to set a scheduled job that purges all tuples once a day and rewrites all tuples form the source of truth db. This will make sure all "dangeling" tuples that might have been missed are purged and all new that might have been missed are added as well. 

However this approach is quite  dirty in my opinion, and i would like to see how others have dealt with changes / synchronizing / patching / migrations in production.

**Reply by bytefish** [reply]
@Sagensagen I have thought about it and it’s a good idea to update my comment. What I have previously described, which was writing to both stores without a shared transaction, is going to lead to the Dual Write Problem. In my example application I have implemented an Outbox Pattern instead:

* https://debezium.io/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/

I think a lot of people are going to use Debezium with an Outbox Pattern to reach Eventual Consistency between the stores. I have used Postgres Logical Replication feature for subscribing to the Outbox table:

* https://www.bytefish.de/blog/outbox_events_postgres_dotnet.html

Actually I would have simply written a periodic job to purge all relations and rewrite all of them at once, so the stack is simple and I don’t need to analyze logs. 

But I couldn’t figure out the OpenFGA APIs to do this efficiently.

**Reply by Sagensagen** [reply]
A possibility to clean the tuples is to directly delete them through psql connection instead of hacking the openfga api somehow. This allows for  DELETE FROM "tuple"; removing all rows. Not sure how i feel about this approach either...
